### ISSUE TITLE
KV Cache for Efficient Conditional Image Generation

### ISSUE THREAD

**Issue by zzzzzero** [Xiaojiu-z/EasyControl#37]
It seems that the open-source code does not achieve the effect described in the paper, where the KV cache is used to save computation by limiting the calculation of condition's key and value.

Currently, key and value are computed over the entire hidden_states:
```
key = attn.to_k(hidden_states)
value = attn.to_v(hidden_states)
```
In fact, it seems that only the features excluding the condition should be used to compute the key and value, like this:
```
key = attn.to_k(hidden_states[:, :scaled_block_size, :])
value = attn.to_v(hidden_states[:, :scaled_block_size, :])
```

**Comment by Xiaojiu-z** [comment]
As illustrated in the figure, the condition branch integrates the pre-trained `attn.to_q,` `attn.to_k,` and `attn.to_v` layers with newly introduced LoRA projections. Notably, these additional LoRA projections exclusively affect the condition branch without interfering with other components. By combining this setup with causal attention during training, we can leverage KV caching to pre-compute the condition features efficiently

<table>
  <tr>
    <td><img src="https://github.com/user-attachments/assets/4c94303f-085f-4d63-a8b9-2d39e9b2a23b" style="width:100%; max-height:250px; object-fit:contain" /></td>
    <td><img src="https://github.com/user-attachments/assets/1e75e83a-4afb-4e46-9350-454369984045" style="width:100%; max-height:250px; object-fit:contain" /></td>
  </tr>
</table>

**Comment by zzzzzero** [comment]
What I mean is, if I understand correctly, the current code re-computes the condition's key and value during each forward pass (which is unnecessary), and then uses the previously cached condition's key and value to replace them.

`norm_hidden_states = torch.concat([norm_hidden_states, norm_cond_hidden_states], dim=-2)`

```
            query = attn.to_q(hidden_states) 
            key = attn.to_k(hidden_states)
            value = attn.to_v(hidden_states)

            inner_dim = query.shape[-1]
            head_dim = inner_dim // attn.heads
            
            query = query.view(batch_size, -1, attn.heads, head_dim).transpose(1, 2)
            key = key.view(batch_size, -1, attn.heads, head_dim).transpose(1, 2)
            value = value.view(batch_size, -1, attn.heads, head_dim).transpose(1, 2)
            
            key = torch.concat([key[:, :, :scaled_block_size, :], self.bank_kv[0]], dim=-2)
            value = torch.concat([value[:, :, :scaled_block_size, :], self.bank_kv[1]], dim=-2)
```

**Comment by Xiaojiu-z** [comment]
Oh I get you. You're absolutely right. While removing redundant linear projections can reduce some computation, the actual time savings during inference are quite limited. Based on my experience, eliminating these duplicate linear layers operation doesn’t lead to significant improvements—the most critical optimization still lies in KV Cache.
